cmd/simulacra/simulacra.go

```diff
@@ -0,0 +1,183 @@
+	return nil
+}
+
+func getAllReceivers(config *confgenerator.UnifiedConfig) (receivers []string) {
+	for _, receiver := range config.Logging.Receivers {
+		receivers = append(receivers, receiver.Type())
+	}
+
+	for _, receiver := range config.Metrics.Receivers {
+		receivers = append(receivers, receiver.Type())
+	}
+	return receivers
+}
+
+// Note: The following functions are mostly a duplicate of helper functions
+// that already exist in gce_testing.go. The reason why we have them here is so that we
+// can use gce.RunRemotely to execute our script. gce.RunScriptRemotely does this for us
+// but it expects a Directory Logger. A directory logger does not make much sense for our
+// purposes.
+
+// envVarMapToBashPrefix converts a map of env variable name to value into a string
+// suitable for passing to bash as a way to set those variables. The environment values
+// are wrapped in quotes. Example output: `VAR1='foo' VAR2='bar' `
+func envVarMapToBashPrefix(env map[string]string) string {
+	var builder strings.Builder
+	for key, value := range env {
+		fmt.Fprintf(&builder, "%s='%s' ", key, value)
+	}
+	return builder.String()
+}
+
+// envVarMapToPowershellPrefix converts a map of env variable name to value into a string
+// suitable for prepending onto a powershell command as a way to set those variables.
+// Example output: "$env:VAR1='foo'\n$env:VAR2='bar'\n"
+func envVarMapToPowershellPrefix(env map[string]string) string {
+	var builder strings.Builder
+	for key, value := range env {
+		fmt.Fprintf(&builder, "$env:%s='%s'\n", key, value)
+	}
+	return builder.String()
+}
+
+func runScriptRemotely(ctx context.Context, logger *log.Logger, vm *gce.VM, scriptContents string, env map[string]string) (_ gce.CommandOutput, err error) {
+	if gce.IsWindows(vm.Platform) {
+		// Use a UUID for the script name in case RunScriptRemotely is being
+		// called concurrently on the same VM.
+		scriptPath := "C:\\" + uuid.NewString() + ".ps1"
+		if err := gce.UploadContent(ctx, logger, vm, strings.NewReader(scriptContents), scriptPath); err != nil {
+			return gce.CommandOutput{}, err
+		}
+		// powershell -File seems to drop certain kinds of errors:
+		// https://stackoverflow.com/a/15779295
+		// In testing, adding $ErrorActionPreference = 'Stop' to the start of each
+		// script seems to work around this completely.
+		return gce.RunRemotely(ctx, logger, vm, "", envVarMapToPowershellPrefix(env)+"powershell -File "+scriptPath)
+	}
+	scriptPath := uuid.NewString() + ".sh"
+	// Write the script contents to <UUID>.sh, then tell bash to execute it with -x
+	// to print each line as it runs.
+	// Use a UUID for the script name in case RunScriptRemotely is being called
+	// concurrently on the same VM.
+	return gce.RunRemotely(ctx, logger, vm, scriptContents, "cat - > "+scriptPath+" && sudo "+envVarMapToBashPrefix(env)+"bash -x "+scriptPath)
+}
+
+// installApps reads an Ops Agent config file and then identifies all the third party apps that need to be installed.
```

Comment:
I think this is pretty neat. Its fine to keep this for now but I wonder if we should have a yaml config where folks can install the apps they care about without the tool inferring it from the Ops Agent config.

This is because there maybe a lot of different parameters to installing an application. The app version, the repo its fetched from, various knobs that the customer environment tweaked etc. Our install and exercise scripts installs the latest app at the time of writing, to validate that the integration works in a simple environment. Since this tool will need to be more flexible, I'm wondering whether we should decouple the installing of the apps from the ops agent config.